File: app/src/main/java/org/emschu/snmp/cockpit/tasks/QueryTaskExecutor.java
package org.emschu.snmp.cockpit.tasks;


import android.os.Handler;
import android.os.Looper;
import android.util.Log;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;

import org.emschu.snmp.cockpit.CockpitStateManager;
import org.emschu.snmp.cockpit.query.QueryCache;
import org.emschu.snmp.cockpit.query.QueryRequest;
import org.emschu.snmp.cockpit.query.SnmpQuery;
import org.emschu.snmp.cockpit.snmp.DeviceConfiguration;
import org.emschu.snmp.cockpit.snmp.SnmpConnection;
import org.emschu.snmp.cockpit.snmp.SnmpManager;
import org.emschu.snmp.cockpit.snmp.model.QueryResponse;

import java.util.List;
import java.util.concurrent.Executor;

public class QueryTaskExecutor {
    private final static String TAG = QueryTaskExecutor.class.getSimpleName();

    private static final Executor executor = SnmpManager.getInstance().getThreadPoolExecutor();
    private static final Handler handler = new Handler(Looper.getMainLooper());

    public interface Callback<R> {
        void onComplete(R result);
    }

    public static <T extends SnmpQuery> void executeAsync(@NonNull QueryRequest<T> queryRequest, @Nullable Callback<T> uiCompletionCallback) {
        executor.execute(() -> {
            //final long startTime = System.currentTimeMillis();
            DeviceConfiguration deviceConfiguration = queryRequest.getDeviceConfiguration();
            if (deviceConfiguration == null) {
                throw new IllegalStateException("null device config!");
            }

            if (queryRequest.isCacheable()) {
                Log.d(TAG, "query has cache id: " + queryRequest.getCacheId());
                QueryCache queryCache = CockpitStateManager.getInstance().getQueryCache();
                T cachedQuery = (T) queryCache.get(queryRequest.getCacheId());
                if (cachedQuery != null) {
                    Log.d(TAG, "return query from cache");
                    // this is executed on the main/ui thread
                    if (uiCompletionCallback != null) {
                        handler.post(() -> {
                            uiCompletionCallback.onComplete(cachedQuery);
                        });
                    }
                    return;
                }
            }

            // get snmp connection instance
            SnmpConnection connector = SnmpManager.getInstance().getOrCreateConnection(deviceConfiguration);
            if (connector == null) {
                Log.d(TAG, "no connection available");
                return;
            }
            if (!connector.canPing(deviceConfiguration)
                    && !CockpitStateManager.getInstance().isInTimeouts()) {
                Log.d(TAG, "reset due to inactivity");
                if (deviceConfiguration.getSnmpVersion() < 3) {
                    SnmpManager.getInstance().resetV1Connection(deviceConfiguration);
                } else {
                    SnmpManager.getInstance().resetV3Connection(deviceConfiguration);
                }
                connector = SnmpManager.getInstance().getConnection(deviceConfiguration);
                if (connector == null) {
                    Log.w(TAG, "no connection available after reset");
                    return;
                }
            }
            connector.startListening();

            List<QueryResponse> queryResponses;
            Log.i(TAG, "querying " + queryRequest.getOidQuery().toDottedString());
            if (queryRequest.isSingleRequest()) {
                queryResponses = connector.querySingle(deviceConfiguration,
                        queryRequest.getOidQuery());
            } else {
                queryResponses = connector.queryWalk(deviceConfiguration,
                        queryRequest.getOidQuery());
            }

            if (queryResponses == null) {
                return;
            }
            T queryResponseClass;
            try {
                queryResponseClass = (T) queryRequest.getQueryClass().newInstance();
            } catch (InstantiationException | IllegalAccessException e) {
                Log.e(TAG, "Error instantiate SnmpQuery class: " + e.getMessage());
                return;
            }
            queryResponseClass.processResult(queryResponses);

            if (queryRequest.isCacheable()) {
                QueryCache queryCache = CockpitStateManager.getInstance().getQueryCache();
                queryCache.put(queryRequest.getCacheId(), queryResponseClass);
            }

            // this is executed on the main/ui thread
            if (uiCompletionCallback != null) {
                handler.post(() -> {
                    uiCompletionCallback.onComplete(queryResponseClass);
                });
            }
        });
    }

}

File: app/src/main/java/org/emschu/snmp/cockpit/tasks/QueryTaskExecutor.kt
/*
 * snmp-cockpit
 *
 * Copyright (C) 2018-2023
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see <https://www.gnu.org/licenses/>.
 */

/*
 * SNMP Cockpit Android App
 *
 * Copyright (C) 2021-22
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see <https://www.gnu.org/licenses/>.
 */

package org.emschu.snmp.cockpit.tasks


import android.util.Log
import org.emschu.snmp.cockpit.SnmpCockpitApp.Companion.cockpitStateManager
import org.emschu.snmp.cockpit.SnmpCockpitApp.Companion.snmpManager
import org.emschu.snmp.cockpit.snmp.DeviceConfiguration
import org.emschu.snmp.cockpit.snmp.OIDCatalog
import org.emschu.snmp.cockpit.snmp.QueryResponse
import org.emschu.snmp.cockpit.snmp.query.SnmpQuery

object QueryTaskExecutor {
    private val TAG = QueryTaskExecutor::class.java.simpleName

    fun querySnmpDevice(
        deviceConfiguration: DeviceConfiguration,
        queryRequest: SnmpQuery,
        isCacheAllowed: Boolean = true,
    ): SnmpQuery? {
        if (isCacheAllowed && queryRequest.isCacheable) {
            Log.d(TAG, "query has cache id: " + queryRequest.cacheId)
            val queryCache = cockpitStateManager.queryCache
            val cachedQuery = queryCache[queryRequest.cacheId + deviceConfiguration.uniqueDeviceId]
            if (cachedQuery != null) {
                Log.d(TAG, "return query from cache")
                return cachedQuery
            }
        }

        // get snmp connection instance
        var connector = snmpManager.getOrCreateConnection(deviceConfiguration)
        if (connector == null) {
            Log.d(TAG, "no connection available")
            return null
        }
        if (!connector.canPing(deviceConfiguration)) {
            if (deviceConfiguration.isV1 || deviceConfiguration.isV2c) {
                snmpManager.resetV1Connection(deviceConfiguration)
            } else {
                snmpManager.resetV3Connection(deviceConfiguration)
            }
            connector = snmpManager.loadConnection(deviceConfiguration)
            if (connector == null) {
                Log.w(TAG, "no connection available after reset")
                return null
            }
        }
        connector.startListening()

        Log.i(TAG, "querying " + queryRequest.oidQuery.toDottedString())
        val queryResponses: List<QueryResponse> = if (queryRequest.isSingleRequest) {
            connector.querySingle(deviceConfiguration, queryRequest.oidQuery)
        } else {
            connector.queryWalk(deviceConfiguration, queryRequest.oidQuery)
        }

        val responseListWithAsn = queryResponses.map { queryResponse ->
            val asnName: String? = OIDCatalog.getAsnByOid(queryResponse.oid)
            if (!asnName.isNullOrBlank()) {
                queryResponse.copy(asnName = asnName)
            } else {
                queryResponse
            }
        }
        queryRequest.processResult(responseListWithAsn)

        if (queryRequest.isCacheable && queryResponses.isNotEmpty()) {
            val queryCache = cockpitStateManager.queryCache
            queryCache.put(queryRequest.cacheId + deviceConfiguration.uniqueDeviceId, queryRequest)
        }

        return queryRequest
    }
}